pkg/csi_driver/multishare_stateful_controller.go

```diff
@@ -0,0 +1,297 @@
+	result, err := m.clientset.MultishareV1alpha1().ShareInfos().Update(ctx, shareInfoClone, metav1.UpdateOptions{})
+	if err != nil {
+		return result, err
+	}
+	return result, nil
+}
+
+func (m *MultishareStatefulController) createShareInfo(ctx context.Context, shareInfo *v1alpha1.ShareInfo) (*v1alpha1.ShareInfo, error) {
+	result, err := m.clientset.MultishareV1alpha1().ShareInfos().Create(ctx, shareInfo, metav1.CreateOptions{})
+	if err != nil {
+		return result, err
+	}
+	return result, nil
+}
+
+func (m *MultishareStatefulController) deleteShareInfo(ctx context.Context, siName string) error {
+	return m.clientset.MultishareV1alpha1().ShareInfos().Delete(ctx, siName, metav1.DeleteOptions{})
+}
+
+func (m *MultishareStatefulController) ControllerExpandVolume(ctx context.Context, req *csi.ControllerExpandVolumeRequest) (*csi.ControllerExpandVolumeResponse, error) {
+	reqBytes, err := getShareRequestCapacity(req.GetCapacityRange(), util.MinShareSizeBytes, util.MaxShareSizeBytes)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	if !util.IsAligned(reqBytes, util.Gb) {
+		return nil, status.Errorf(codes.InvalidArgument, "requested size(bytes) %d is not a multiple of 1GiB", reqBytes)
+	}
+	_, _, _, _, shareName, err := parseMultishareVolId(req.VolumeId)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	klog.Infof("ControllerExpandVolume called for multishare with request %+v", req)
+
+	siName := util.ShareToShareInfoName(shareName)
+	shareInfo, err := m.shareLister.Get(siName)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, status.Errorf(codes.Internal, "error getting shareInfo %q from informer: %s", siName, err.Error())
+		}
+		klog.Infof("shareInfo %s does not exist in cache", siName)
+		return nil, status.Errorf(codes.Aborted, "waiting to express intent for volume %s to be expanded", siName)
+	}
+
+	if shareInfo.Spec.CapacityBytes < reqBytes {
+		// update Spec.CapacityBytes
+		shareInfoClone := shareInfo.DeepCopy()
+		shareInfoClone.Spec.CapacityBytes = reqBytes
+		shareInfo, err = m.updateShareInfo(ctx, shareInfoClone)
+		if err != nil {
+			klog.Errorf("failed to update shareInfo %s: %s", siName, err.Error())
+			return nil, status.Errorf(codes.Internal, "error expanding volume %s due to failed internal update", siName)
+		}
+		return nil, status.Errorf(codes.Aborted, "expressed intent for volume %s to be expanded", siName)
+	}
+
+	if shareInfo.Status == nil {
+		klog.Errorf("ControllerExpandVolume called for %s but shareInfo.Status is nil", siName)
+		return nil, status.Errorf(codes.Internal, "volume %s is not yet created", siName)
+	}
+
+	if shareInfo.Status.CapacityBytes >= reqBytes && shareInfo.Status.ShareStatus == v1alpha1.READY {
+		klog.Infof("Controller expand volume succeeded for volume %v, size(bytes): %v", req.VolumeId, shareInfo.Status.CapacityBytes)
```

Comment:
we should also verify with getshare() the real share capacity just to be sure (https://github.com/kubernetes-sigs/gcp-filestore-csi-driver/blob/master/pkg/csi_driver/multishare_controller.go#L352)